extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/kafkastream/KafkaStreamsMetricsProducer.java

```diff
@@ -0,0 +1,41 @@
+package io.quarkus.micrometer.runtime.binder.kafkastream;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.Produces;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.apache.kafka.streams.KafkaStreams;
+
+import io.micrometer.core.instrument.binder.kafka.KafkaStreamsMetrics;
+import io.quarkus.runtime.ShutdownEvent;
+import io.quarkus.runtime.Startup;
+
+@ApplicationScoped
+public class KafkaStreamsMetricsProducer {
+
+    KafkaStreamsMetrics kafkaStreamsMetrics;
+
+    @Inject
+    public KafkaStreamsMetricsProducer(Instance<KafkaStreams> kafkaStreams) {
+        if (!kafkaStreams.isUnsatisfied()) {
+            kafkaStreamsMetrics = new KafkaStreamsMetrics(kafkaStreams.get());
+        }
+    }
+
+    @Produces
+    @Singleton
+    @Startup
```

Comment:
@Startup is also unnecessary, because the producer will be driven by MeterBinder discovery (which is essentially at startup).

I worry a little bit about initialization dependencies.. as resolving the MeterBinder (which is what will trigger this bean to be produced) then requires kafka streams to be started/resolved, which should be fine, but.. 

I wonder if we should do something a little more like what the rest of Kafka metrics is doing, so that you have this @Startup bean which waits for KafkaStreams to be created / injected into the constructor.. 

When that happens, it registers Kafka metrics with the global meter registry.
And then the onStop method is unchanged?

Reply:
I will take a closer look at the producer of KafkaStreamsMetrics.

Because of the scheduler in KafkaMetric, the dependency on the state of KafkaStream should not be that delicate.

But I will see if I can find a more elegant solution.
Maybe I can do the coupling similar to smallrye-reactive-messaging via a CDI event.

But kafkaStreams is created in the constructor and I don't know how well events work there. Tomorrow I will see if this works or if I see another possibility. Another possibility is to publish the event in @PostConstruct.

KafkaStreams creation: https://github.com/quarkusio/quarkus/blob/9d3ebaf418a7d9f378f60b1032944f3f4b095ec2/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java#L85

Reply:
I changed the logic and adapted it to the existing KafkaEventObserver. For this I had to make adjustments to `io.quarkus.kafka.streams.runtime.KafkaStreamsProducer`.

@gunnarmorling can you also look at the changes? Not sure if I should publish the CDI event in the constructor or if the PostConstruct is ok.

Reply:
Or @cescoffier  ?

Reply:
I think the current approach is fine, firing that event from post-construct.